text/0000-confidential-token-ids.md

```diff
@@ -0,0 +1,270 @@
+- Feature Name: `confidential_token_ids`
+- Start Date: 2022-02-05
+- MCIP PR: [mobilecoinfoundation/mcips#0025](https://github.com/mobilecoinfoundation/mcips/pull/0025)
+- MobileCoin Epic: TBD
+
+# Summary
+[summary]: #summary
+
+Enable new token types to be added to the protocol and transacted on chain without revealing or leaving a transparent record of the token type.
+
+# Motivation
+[motivation]: #motivation
+
+The MobileCoin ledger protocol currently supports one token type, MOB. We would like to allow the addition of assets to the ledger without revealing which assets were involved in any transaction. The transaction confidentiality and integrity guarantees should be the same for the MobileCoin blockchain no matter how many token types are supported. Namely, the sender, the recipient, and the amount should remain private to all except the participants in the transaction, and no double spending or unauthorized minting should be possible. In addition, we now specify that the asset type should remain private to all except the participants in the transaction.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A Transaction Output (TXO) has a confidential TokenType enum field (`masked_token_id`). Transactions with inputs and outputs, including a fee output, must be composed of a single TokenType. Some tokens may have additional functionality, such as minting and burning, which require verification of the TokenType as part of the authorization of the action.
+
+Each TokenType has an explicit minimum fee specified by node operators, and included in the attestation handshake, ensuring that all nodes in the network are configured with the same fee minimums and TokenType sets. New TokenTypes can be added only with unanimous agreement from all node operators. 
+
+A transaction specifies the TokenType for the entire transaction, and includes a "Proof of Opening" that all `token_ids` in the inputs and outputs are the same. This prevents spending one TokenType as a member of another TokenType. The TransactionBuilder now takes a `token_id` as an argument in its initializer. The TokenType is set in the clear on the TxPrefix for a transaction, which is only accessed in the enclave, after the payload is delivered via an encrypted, attested channel. This is similar to the TxInputs, and thus has a similar threat model for confidentiality. See [Confidentiality Analysis](#confidentiality-analysis) for an assessment of risk.
+
+Clients must choose how to expose multiple asset types to their users. There may be clients who wish to only expose a subset of asset types, or clients who wish to provide balances and transactionality in all asset types. 
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
```

Comment:
I would consider adding a section here calling out explicitly how `Amount` has changed:

Previously, an amount of picomob in a TxOut is represented by a Pedersen commitment to that number, as well as a "masked_value", which is created by XORing the number of picomob with a pseudorandom mask (derived by hashing a shared secret). The purpose of the `masked_value` is that when the recipient receives a TxOut, they can unmask this value, then reconstruct the Pedersen commitment from the number, to confirm successful recovery of the value.

After this change:
* A `masked_token_id` is added, which, similarly to masked_value, provides a way for the recipient to learn the token id (and then subsequently confirm it).
* The pedersen commitment to the value now is made using a basepoint `H_i` which depends on the token id. (The basepoint for the blinding factor is the same as before regardless of the token id.)
* Thus, the recipient can attempt to scan a TxOut by first constructing the shared secret, then unmasking the masked_amount and masked_token_id, and then reconstructing the Pedersen commitment. If this reconstruction is successful, then the recipient knows the amount and the token id, similarly as before the token ids.

### Pedersen generators

The requirements for the pedersen generators are:
* The generator used for MOB is the same as before, so `H_0` is a known value.
* The set of all generators used for token ids should be cryptographically "orthogonal", meaning that, it should be intractable for anyone to find a linear relationship between them.
* The function mapping a token id `i` to the generator `H_i` must be implemented in constant-time.

The established value for `H_0` in mobilecoin is to take the `RISTRETTO_BASEPOINT` bytes, hash them using blake2b, and then hash this to the ristretto curve using the elligator map exposed by curve25519-dalek.
We propose to compute `H_i` by converting `i` to little endian bytes, and XOR'ing these bytes over the first four bytes of the `RISTRETTO_BASEPOINT` bytes before hashing them with blake2b, because this is a simple way to meet these requirements.

Reply:
> This would seem to imply that, as additional new token types are added, that each token type's basepoint must be tried separately to recover the opening to the commitment. Is that accurate or did I misunderstand something?

No, that's not the proposal. What you do is, you unmask the `masked_token_id`, which is based on the assumption that this `TxOut` belongs to you, and so your shared secret computation is correct, so that you can compute the token id mask. Then you get a token id which is the only one it can be if it belongs to you, so you only have to test one value.

Reply:
> This might in practise be fine, but I do not think it meets the standard cryptographic definition of orthogonal, since it depends at each stage upon the original Ristretto basepoint in compressed form.

I guess I don't agree: what it depends on is the assumption that `i -> Blake2b( RistrettoBasePoint XOR i)` is hard to distinguish from a random function in some sense. I would be happy with this assumption for any particular bytes in place of `RistrettoBasePoint`, just as I am happy with any particular domain separator string when hashing. Those bytes are only there because one of the requirements is that the value of this function for `i = 0` matches what we are using for MOB right now.

It seems like `i -> Blake2b( RistrettoBasePoint XOR i)` is just as good of a "nothing up my sleeve sequence" as digits of pi, they are both random arbitrary strings that are hard to control. If we went with digits of pi then this would no longer be efficient to compute for large values of `i`, and it needs to be efficient to compute for large values of `i`.

> (Also the Ristretto flavour of the Elligator2 mapping is not meant to be publicly exposed; is this something that was changed in a fork?)

I misread the docu for `RistrettoPoint::from_hash`, I guess it's actually taking a 64 byte hash and then doing what you said.

https://docs.rs/curve25519-dalek/latest/src/curve25519_dalek/ristretto.rs.html#704-713

Reply:
Incorporating the text from @garbageslam's initial suggestion - should I incorporate any changes based on the discussion in this point, @isis-mc?

Reply:
Incorporating the text from @garbageslam's initial suggestion - should I incorporate any changes based on the discussion in this point, @isis-mc?